### ISSUE TITLE
User experience of connecting with websites

### ISSUE THREAD

**Issue by graup** [aergoio/wallet-extension#6]
Websites should be able to request a user-installed wallet (either this extension or a thrid-party, application-specific extension) to sign transactions. How to give access to an Aergo provider to a dapp? There has been plenty of discussion about the UX of this in the Ethereum space. Some considerations are ease of use, but also privacy (not letting websites fingerprint users as crypto users). This issue acts as a place for discussing ideas. We need to make a decision of what's important in Aergo.

In Ethereum, Metamask or Mist traditionally provided a complete web3 object so that a website can interface with the included full node (or another remote node). But the main requirement here is just to give access to signing using user data inside the wallet. Everything else that might be necessary can be better handled by the website itself. So I propose to only offer the absolutely necessary methods, but not a full-fledged provider like web3 does. 

Nevertheless, we can learn a lot from previous discussions, esp. regarding UX, security, and privacy.

References
- https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1102.md
- https://ethereum-magicians.org/t/the-ux-of-eip-1102/814/25?u=bitpshr
- https://ethereum-magicians.org/t/eip-1102-opt-in-provider-access/414
- https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md

Human verifiable data signing
- https://eips.ethereum.org/EIPS/eip-712

**Comment by graup** [comment]
I’m planning to start with a simple interface like this.

1. Website sends request using `[postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage)` or similar.
2. Aergo Connect opens popup showing parameters and asking to confirm
3. Result is passed back to website through event

The reason for this generic asynchronous protocol is that the extension WILL NOT expose itself to the website (like Metamask used to do by default, i.e. inject a `window.web3` object) because that has privacy implications. (Any website can find out that you are an Aergo user.)

I suggest 3 request types for now:

1. sign arbitrary message
2. send tx
3. get active account

Example code:

```
// send request to aergo connect
window.postMessage({ type: 'aergo_connect_sign', message: "0x12321e1efee1" }, '*');
// or
window.postMessage({ type: 'aergo_connect_sendtx', tx: { amount: 123, recipient: "Am...." }, '*');

// receive result
window.addEventListener('aergo_connect_sign_result', (result) => {
    // do something with result.signature...
});
// or
window.addEventListener('aergo_connect_sendtx_result', (result) => {
    // do something with result.hash...
});
```

**Comment by paouvrard** [comment]
What about transactions that execute contract functions : users should see the contract address, function name and parameters before approving tx - is that part of 'aergo_connect_sendtx' ?
And also supporting contract deployment would be useful imo.

**Comment by graup** [comment]
Yes. You'd just pass the tx params and they would be displayed in the wallet before signing. Call data is just json thankfully, so it's easy to make it human readable.

You can do contract deployments, but the payload will be not human readable.

**Comment by graup** [comment]
I have a demo ready in this branch: https://github.com/aergoio/wallet-extension/tree/api Build the project, install the dev version, and then go to this demo page:  https://aergoio.github.io/wallet-api-test/

Currently implemented:
- Get active account
- Sign message

Todo:
- Sign tx
- Sign+send tx

**Comment by graup** [comment]
Bugs until now have been fixed, seems to work ok. At the moment this uses browser native `confirm`s, but we should really have a custom UI like Metamask. Need to open a popup with a slightly adjusted UI.

**Comment by graup** [comment]
I added a proper UI using a popup. Currently writing documentation, should be released together with Aergo 2.0 until late November.

**Comment by graup** [comment]
This is now live and deployed on the chrome webstore as version 0.2.0.

Docs: https://github.com/aergoio/wallet-extension/blob/master/API.md
